### PR TITLE
Revert signature back to v3.

### PR DESCRIPTION
**Pull request by kat-co**
Some users of goamz will not be able to pass in an io.ReadSeeker. Reverted the signature of PutReaderWithHeader to take in an io.Reader. The reader is then converted to an io.ReadSeeker through either a type assertion, or by reading the entire contents into memory. This allows for an optimization to those who pass in an io.ReadSeeker, but allows use for everyone.

### CODE REVIEW

File: s3/s3.go [hunk 1]
```diff
@@ -236,7 +245,12 @@ func (b *Bucket) PutReaderWithHeader(path string, r io.ReadSeeker, length int64,
 		return err
 	}
 	// Signing may read the request body.
-	if _, err := r.Seek(0, 0); err != nil {
+	if _, err := seeker.Seek(0, 0); err != nil {
```
Comment (axw) [inline s3/s3.go]: Should seek back to the position prior to signing, which is not necessarily the beginning.

File: s3/s3.go [hunk 2]
```diff
@@ -236,7 +245,12 @@ func (b *Bucket) PutReaderWithHeader(path string, r io.ReadSeeker, length int64,
...
+		return err
+	}
+
+	// Signing may read the request body.
+	if _, err := seeker.Seek(0, 0); err != nil {
```
Comment (axw) [inline s3/s3.go]: why twice?

**Comment by axw** [comment]
LGTM with fix

File: s3/s3.go (hunk 1 above)
Comment (kat-co) [inline s3/s3.go]: Good point; fixed.

File: s3/s3.go (hunk 2 above)
Comment (kat-co) [inline s3/s3.go]: Held down the paste keys too long? :) Corrected.

File: s3/s3.go [hunk 3]
```diff
@@ -205,21 +205,30 @@ func (b *Bucket) GetReader(path string) (rc io.ReadCloser, err error) {
...
+	seeker, ok := r.(io.ReadSeeker)
+	if !ok {
+		content, err := ioutil.ReadAll(r)
+		if err != nil {
+			return err
+		}
+		seeker = bytes.NewReader(content)
+	}
 
-	req, err := http.NewRequest("PUT", b.Region.ResolveS3BucketEndpoint(b.Name), r)
```
Comment (nathany) [inline s3/s3.go:222]: Does using seeker here still work when `r` is not a ReadSeeker?

**Pull request merged**